feat: Add new process and move contributors

BekahHW opened the pull request:
<!-- Please fill in all areas in this PR form. Incomplete PRs will be marked invalid and may be closed. -->

## Description

<!--
Please do not leave this blank. Add your description **below** this line and **outside** of the comment tags.
For example: This PR adds <your-github-username> as a contributor.
-->

This pull request introduces a new, automated, and conflict-free contributor management workflow for the guestbook repository. It adds contributor JSON files for multiple users, updates the contribution process and documentation, and implements several GitHub Actions to validate, welcome, and update contributors efficiently. The changes are grouped below by theme.

**Automation and Workflow Enhancements:**

* Added `.github/workflows/validate-pr.yml` to automatically validate contributor PRs, ensuring only one contributor file is added per PR, the file is correctly named, and required fields are present. It also welcomes new contributors upon PR creation.
* Introduced `.github/workflows/validate-contributors.yml` to validate contributor JSON files for structure, duplicates, and syntax on PRs and pushes.
* Added `.github/workflows/update-contributors.yml` to automatically update the contributors section in the README using the `all-contributors` CLI whenever contributor files are added or changed.
* Created `.github/workflows/welcome-new-contributor.yml` to post a congratulatory comment when a contributor PR is merged.

**Contributor Onboarding and Documentation:**

* Updated `CONTRIBUTING.md` to direct users to a new step-by-step guide at `docs/guides/contributor-guide.md`, clarified the process for adding oneself as a contributor, and added guidance for other contribution types. [[1]](diffhunk://#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055L3-R10) [[2]](diffhunk://#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055L23-R35)
* Overhauled `.github/PULL_REQUEST_TEMPLATE.md` to streamline the PR process, clearly separate new contributor submissions from other PR types, and provide explicit checklists and instructions.

**Contributor Data Management:**

* Added a `.gitkeep` file to ensure the `contributors` directory is tracked by git and included initial contributor JSON files for several users (e.g., `02zeda.json`, `ATREAY.json`, etc.). [[1]](diffhunk://#diff-e667983a70a36e14cff4bed25ffe9644fae8bf2706293a194cd37054c1a1fba4R1-R2) [[2]](diffhunk://#diff-f48ca6e034f986170b2128d3ae9bddb4db790cfac1f7e4919fb172e218ca1469R1-R11) [[3]](diffhunk://#diff-df986cb1bddc902fca13a50f074408ebd265905e84dc2aa75e2de366f17442d4R1-R17) [[4]](diffhunk://#diff-c223bb8973749afc78d11e97b3c1870df0e392dc438ff9579a6d569e8c0c9286R1-R40) [[5]](diffhunk://#diff-699a681b55ec1b1cda0702afdc87750aad8ec51a6b61759f07ae24eccc31d040R1-R13) [[6]](diffhunk://#diff-2127a14d43ef0af3e4accc5740f012188a5eb5a9afb5e80d38c8386d07f42026R1-R13) [[7]](diffhunk://#diff-95c5caa4d9f712793f8fc7897f11e196f9c164fc8634214a3e2ee00dee079078R1-R14) [[8]](diffhunk://#diff-3abc53bbd4a95cb0135d76bb8a0d56a8a1be973fa8cd7956446430363b309721R1-R8)

**Minor Improvements:**

* Updated the post-PR comment in `intro_course_contributor.yml` to streamline messaging for course contributors.

These changes collectively modernize and automate the contributor onboarding process, reduce merge conflicts, and improve the contributor experience.

## What type of PR is this? (check all applicable)

- rehaul the process

## Related Issues

<!--
Add your related issue **below** this line and **outside** of the comment tags.



Please use this format to link your issue: Closes #XXX.
Change "XXX" to your issue number that you can find next to your issue's title.

More information about link issue: https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue
-->

closes #827 

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🙅 no documentation needed

## Screenshot (Required for PR Review)

<!-- Please provide a screenshot of your profile being generated on the README. This ensures that you ran the `npm run contributors:generate` command, as mentioned in the previous question, which makes it easier for the maintainers to review PRs. All PRs without screenshots will be automatically rejected-->

## [optional] What GIF best describes this PR or how it makes you feel?

<!-- note: PRs with deleted sections will be marked invalid -->

<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.

  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Contributing Guide: https://github.com/OpenSource-Communities/guestbook/blob/main/CONTRIBUTING.md.
  - 📖 Read the Code of Conduct: https://github.com/OpenSource-Communities/.github/blob/main/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

BekahHW committed:
Add new process and move contributors

BekahHW commented:
@adiati98 I still need to test this again, but I _think_ this is working. The majority of the files changed are moving the contributors to their own json file. The docs/MiGRATION_GUIDE.md explains the changes that have been made. 

I wanted to give you a chance to read through everything and ask questions.

adiati98 commented:
@BekahHW thanks so much for this!

I'll test this locally and get back to you.

adiati98 commented:
@BekahHW I've tested this out locally and here's my thoughts:

## Missing Instructions

### npm install

There's no instruction anywhere to run `npm install`. For folks to be able to run any npm command, they need to install it first.

#### Suggestion

Add an instruction to run `npm install` as the first step to do before creating a json file in:

- "Quick Start" on root README 
- "Step-by-Step Instructions" on `docs/guides/contributor-guide.md` 

### Add, commit, and push changes

Once a contributor added their JSON file, there's no instruction to add, commit, and push their changes to create a PR. For new folks in OSS, they would get confuse of what to do next after adding their JSON file.

These instructions, however, are available on the terminal **only** if they chose to preview their profile as stated in the instruction: "5. **Test locally (optional)**" at `docs/guides/contributor-guide.md`.

#### Suggestions

There are a couple of options:

1. Add instructions to add, commit, and push changes before test locally
2. Make instruction no. 5 above as non optional, so folks can preview their profile before add, commit, and push, then follow the instructions from the terminal to push their changes.

I personally recommend no. 2 here.

### Push changes

After running the **test** command — `npm run contributors:test USERNAME` — there will be a next steps message as below:

```bash
🚀 Next steps:
  1. Commit your file: git add contributors/USERNAME.json
  2. Create your PR: git commit -m 'Add USERNAME as a contributor'
  3. Submit and wait for merge!
  4. Check the live README in 5 minutes after merge
```

For folks who are new to OSS, they wouldn't know that "submit" in step 3 is to run the `git push` command.

### Suggestions

Change the instructions to be the same with when running the **preview** command — `npm run contributors:preview USERNAME` — which has clearer steps:

```bash
🚀 Next steps:
  1. git add contributors/USERNAME.json
  2. git commit -m 'Add USERNAME as a contributor'
  3. git push origin your-branch-name
  4. Create your pull request!
```

### Profile Details

In the "Step-by-Step Instructions" section at `docs/guides/contributor-guide.md`, instruction no. 2 is as follow:

````markdown
2. **Add your information**
   Copy this template and fill in your details:

   ```json
   {
     "name": "Your Full Name",
     "github": "your-github-username",
     "profile": "https://your-website.com",
     "contributions": ["code", "doc", "ideas"]
   }
   ```
````

We have info about available contribution types in step 3. But we don't have info what to fill for the `"profile": "https://your-website.com"`. The example here is using personal website URL. But for folks who don't have personal website, there would be questions about what to fill here.

#### Suggestions

Add clear info what should be filled here. This can either their personal website or URL of their GitHub profile.

---

## Running npm Commands

The scripts are written in Python. I don't install Python on Windows. I believe some contributors also don't have Python installed on their computer.

So, for example, when I ran `npm run contributors:preview adiati98`, I got this error:

<img width="1777" height="594" alt="Python was not found error message on terminal" src="https://github.com/user-attachments/assets/0061e6a6-6f26-4b67-930a-4ff8e6c4e54c" />

### Suggestions

We can either:

- Add an instruction to install Python 3, or
- Translate the scripts to JavaScript

---

## Profiles on README < Contributor JSON files

Based on the below from the "Implementation Status" at `docs/MIGRATION_GUIDE.md`:

```markdown
**README complete** with all 310 contributors displaying correctly
```

Before testing anything, I checked the profiles on README and compared them with the total JSON files in the contributors folder.

Here's what I found:

-  ✅ Badge says **310**
-  ❌  There are **only 121** profiles generated on README
-  ❌  There are **311** JSON files in the `contributors` folder — excluding `.gitkeep` and `README.md`.

The contributors' JSON files are much more than those generated on README. So, **README is not complete**.

My wild guess here, because the README will be generated with GH Action, there's a flaw or something missing in the update-contributors workflow: `.github/workflows/update-contributors.yml`.

### Notes

FYI, there's a conflict that needs to be merged as I'm writing this. I've pulled it out to test.

- Once it's pulled, there are merge conflicts in the README
- Clicked "Accept Incoming Changes" as the conflicts are duplication. The incoming changes hold the accurate data. Both badge and total contributors are 311.
- Ran `npm run contributors:migrate`
- It added 2 new JSON files:
  - A contributor that most probably got merged when you were still working on this PR
  - I think I remember this one. It was a PR that's adding two different GH profiles. One of them that got pulled in this testing process was `Added ManeDev.json` — I've checked this one and it goes to 404. So, we can safely remove it.
  
  Removing the 2nd one above should give us 310 total contributors.  But I can't test this out until this PR is merged.

---

## Test Command

Running `npm run contributors:test USERNAME` command won't give us the preview of user profile as if we run the preview command. Below is the screenshot:

<img width="1576" height="798" alt="Could not extract contributors section error on terminal" src="https://github.com/user-attachments/assets/a1d45313-bf16-4f30-be10-09b90dd385f0" />
 
---

## Add and Generate Profile Commands

This is just a reminder. If we will apply all changes (new workflows) in this PR, we need to remove these from `scripts` in the `package.json`:

```json
    "contributors:generate": "all-contributors generate",
    "contributors:add": "all-contributors add",
```

### Thoughts

This other possibility thought just cross my mind. I'm thinking — if we want to keep everything as easy/simple as possible and only if it's possible — we can keep the ability of `contributors:add` command. Maybe add a new script for this instead of using the all-contributors like the current script.

Running this command would:

- Prompt contributors to fill in their details.
- Once it's done, it will automatically generate a new JSON file with their username as file name and store it in the contributions folder.
- Unlike the current command that will automatically add and commit their changes, they still need to add, commit, and push their changes manually because these are crucial steps for contributing to any OSS project. That way, they can experience and practice this flow themselves.

The downside with this compare to manually add their profile that I can think of is probably if folks add their profile manually, then can learn to follow instructions, step by step. In my opinion adding things manually is a great learning process than to automate (almost) everything.

 ---

I might've missed something to test here, but I want to let you know what I found so far.

I've read through everything here, but I want to read one more time. Next, I will have another comment for the GH Actions workflow based on my read and understanding.

adiati98 reviewed (commented):
@BekahHW, I have some questions in the comments.

Other than these, I only have one question that I can think of atm.

In the MIGRATION_GUIDE.md:
> Zero merge conflicts, even with simultaneous contributions.

I just want to confirm my understanding here. So, a contributor profile will be generated on the README after we merge the PR. Is this why there won't be any merge conflicts? Even when there are multiple PRs that we merge, say, just in a minute difference?

Is there anything else that you want me to pay more attention to and give some feedback?

I'll get back to you whenever I have other things in mind.

Thank you! ✨

BekahHW commented:
Yes. You should be able to merge multiple PRs quickly with no issue. It's set up like the VC Hacktoberfest issue.

BekahHW committed:
Update .github/PULL_REQUEST_TEMPLATE.md

Co-authored-by: Ayu Adiati <45172775+adiati98@users.noreply.github.com>

BekahHW committed:
Update .github/workflows/validate-pr.yml

Co-authored-by: Ayu Adiati <45172775+adiati98@users.noreply.github.com>

BekahHW committed:
Add node.js testing and address some comments

BekahHW committed:
Add node.js testing and address some comments

BekahHW committed:
Fix merge errors

BekahHW committed:
Change discussion link in CONTRIBUTING.mdfix: update discussion link in CONTRIBUTING.md to correct repo

Updated discussion link for seeking help.

BekahHW committed:
Update contribution sharing link in workflowfix: replace OpenSauced Discord link with GitHub Discussions in welcome workflow

BekahHW committed:
fix: remove defunct OpenSauced links and update community reference in validate-pr

Updated the PR validation message to enhance clarity and encourage continued contributions.

BekahHW committed:
fix: remove defunct OpenSauced links from testing guide next steps

Updated next steps to focus on continuing contributions to open source.

BekahHW committed:
Update pull request template for clarityfix: convert About Me fields to list format and remove defunct OpenSauced link

BekahHW committed:
fix: replace invalid all-contributors CLI flags with direct .all-contributorsrc update

Refactor contributor addition logic to use Python script for updating .all-contributorsrc.

BekahHW committed:
Merge branch 'main' into feat/contributor-pr-help

BekahHW committed:
fix: replace Python3 with Node.js in update-contributors workflow

BekahHW committed:
fix: correct YAML indentation in update-contributors workflow

BekahHW committed:
fix: correct Python to Node.js translation with proper YAML indentation

BekahHW committed:
Merge branch 'main' into feat/contributor-pr-help
